### PR TITLE
Lift kernel registry to libreyolo/kernels; opt-in hub MSDA kernel for DETR families

### PR DESCRIPTION
**Pull request by EHxuban11**
Opened by an agent. Verified: full CPU unit runs listed below. Not verified: the CUDA hub-kernel path (this machine has CPU-only torch).

## What changed

- Kernel registry lifted from `libreyolo/quant/kernels/` to library-wide `libreyolo/kernels/`, organized by purpose:
  - `kernels/quant/simulate/`: fake-quant Triton kernels (any device, STE backward; serve QAT and simulated PTQ/val).
  - `kernels/quant/execute/`: finalized-only real-precision paths (`scaled_mm_fp8`, fused fp8 cast/epilogue split out of `fp8.py` into `fp8_fusion.py`, `unpack_*`).
  - `kernels/attention/`: new `ms_deform_attn` op slot.
- New accelerated multi-scale deformable attention: with `pip install libreyolo[hub-kernels]`, the compiled CUDA MSDA kernel from Hub repo `kernels-community/deformable-detr` (Apache-2.0) serves RF-DETR, LibreDeformableDETR, and LibreDINO-DETR, forward and backward. Installing the extra is the opt-in (no package, no change, no network access); `LIBREYOLO_HUB_KERNELS=0` disables it. Eligible inputs are eager CUDA fp32; tracing/ONNX/torch.export always use the portable grid_sample path, so exports are untouched; load or runtime failures disable the provider for the process and fall back with one warning.
- `LIBREYOLO_KERNELS` supersedes `LIBREYOLO_QUANT_KERNELS` (legacy alias still honored). `libreyolo.quant.kernels` stays a working alias via a lazy shim.
- Docs: new `docs/kernels.md` (incl. per-platform behavior: CPU/MPS fall back to reference ops, ROCm can engage Triton); `docs/quantization.md` pointers updated. CHANGELOG and THIRD_PARTY_NOTICES updated.
- Why shared-code change: the slot is consumed by three families; the registry mechanism already existed and only moved.

## Reviewer notes

- The per-family MSDA ports stay byte-equivalent fallbacks; call sites only gained a guarded slot lookup that resolves to None by default.
- Import cycle `libreyolo.kernels` <-> `libreyolo.quant` is handled by a PEP 562 shim; `tests/unit/kernels/test_registry.py` covers the kernels-first import order in a subprocess.
- Not verified: actual Hub kernel execution on CUDA (needs a CUDA box with the `kernels` package; predicate-gated, worst case falls back with a warning). Suggested check: install the extra, RF-DETR predict + short train step on GPU, compare against `LIBREYOLO_HUB_KERNELS=0`.

## Tests

- `tests/unit/kernels` + `tests/unit/test_quantize.py`: 66 passed, 22 CUDA-only skipped (CPU box). New tests cover the registry move, env aliases, the import cycle, the MSDA slot default/opt-out, and the call-site layout adapters (numeric equivalence on CPU).
- Deformable-DETR / DINO-DETR / RF-DETR suites (inference, parity, export, registry, postprocess, pose/obb/preprocess, trainer smokes): all passed.
- Full unit tier collection: 5017 tests collect with no import errors.

## Code provenance

- Moved files are pre-existing LibreYOLO code (git renames; MIT).
- New code (registry generalization, `kernels/attention/ms_deform_attn.py`, shim, tests, docs) was written for this PR.
- The autograd bridge in `kernels/attention/ms_deform_attn.py` follows the `MSDeformAttnFunction` interface of Deformable-DETR (https://github.com/fundamentalvision/Deformable-DETR, commit 11169a60c33333af00a4849f1808023eba96a931, Apache-2.0, Copyright (c) 2020 SenseTime). The compiled kernel itself is NOT vendored: it is fetched at runtime, only when the optional `kernels` package is installed, from the Apache-2.0 Hub repository kernels-community/deformable-detr. THIRD_PARTY_NOTICES.txt updated accordingly.

<details><summary><h3>Greptile Summary</h3></summary>

The PR moves the quantization kernel registry into a library-wide hierarchy and adds opt-in accelerated multi-scale deformable attention for DETR-family models. The follow-up pins the fetched native Hub kernel to an immutable revision, addressing the prior supply-chain concern.
- Reorganizes simulation and execution kernels under `libreyolo/kernels/`.
- Preserves the former quantization-kernel import path through a lazy compatibility shim.
- Adds a guarded Hub-backed CUDA attention implementation with portable fallbacks for unsupported inputs and exports.
- Pins the Hub artifact to a full commit revision and adds CUDA parity coverage for future revision changes.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge.

No blocking failure remains; the native Hub artifact is now requested with a full commit revision at the only `get_kernel` call site, fixing the previously reported mutable-artifact path.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| libreyolo/kernels/attention/ms_deform_attn.py | Adds the guarded Hub-backed CUDA attention provider and now loads its native artifact using a full immutable commit revision. |
| libreyolo/kernels/__init__.py | Generalizes the kernel registry and lazily imports a closed list of in-tree providers. |
| libreyolo/models/deformable_detr/ms_deform_attn.py | Consults the accelerated attention slot before retaining the existing portable implementation as fallback. |
| libreyolo/models/rfdetr/transformer.py | Adds the accelerated slot before RF-DETR's layout conversion while preserving the export-compatible portable path. |
| libreyolo/quant/kernels.py | Provides a lazy compatibility shim for the registry's former public import path. |
| pyproject.toml | Adds the optional Hub-kernel runtime loader extra. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    Caller[DETR-family attention caller] --> Guard{Exporting or unsupported input?}
    Guard -->|Yes| Portable[Portable grid_sample implementation]
    Guard -->|No| Registry[Resolve ms_deform_attn slot]
    Registry --> Available{Pinned Hub provider available?}
    Available -->|No| Portable
    Available -->|Yes| Native[Execute pinned CUDA forward/backward]
    Native --> Success{Execution succeeds?}
    Success -->|Yes| Output[Return accelerated result]
    Success -->|No| Disable[Disable provider and warn once]
    Disable --> Portable
```
</details>

<sub>Reviews (3): Last reviewed commit: ["Fix RF-DETR slot wiring, pin the Hub ker..."](https://github.com/libreyolo/libreyolo/commit/e7ceee9c3d1022536d653ba35027af2c1f236699) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=49808431)</sub>

### CODE REVIEW

**Comment by EHxuban11** [comment]
Pushed e7ceee9c addressing review findings:

**Bug: the RF-DETR wiring was unreachable.** The slot consult sat in `ms_deform_attn_core_pytorch` behind `value_spatial_shapes_hw is None`, but RF-DETR threads the hw pairs through its decoder unconditionally (`Transformer.forward` -> decoder layers -> `forward_post`), so no eager forward could ever hit the kernel. The consult now lives in `MSDeformAttn.forward`, gated on `not self._export`, placed *before* the portable path''s transpose — `value` is still `(bs, Len_in, d_model)` there, one zero-copy `view` from the slot''s layout, so the fast path also drops the old adapter''s `permute + contiguous` double copy. The new regression test (`test_rfdetr_attention_forward_routes_through_slot`) goes through the real attention module with `input_spatial_shapes_hw` supplied, exactly as the model calls it.

**Supply chain (Greptile''s finding): Hub revision pinned.** `get_kernel` now fetches `kernels-community/deformable-detr` at commit `4d2393e5d7879f7cf68db04cc7c9c7342272bc05`; a moved branch can no longer swap the native binary. Newer `kernels` releases (0.12+) require a pin anyway.

Also: guard against a first-time kernel fetch inside CUDA graph capture (warmup loads it; capture never does I/O), `torch.compiler.is_exporting` added to the export bypass for non-strict `torch.export`, zero-batch inputs rejected in `_supported_inputs`, and a CUDA-only forward+backward parity test (`test_hub_matches_portable_on_cuda`) that is the designated GPU smoke for revision bumps.

Verified on this (CPU) box: `tests/unit/kernels` 14 passed / 19 CUDA-skips; rfdetr suites 198 passed / 21 skipped; deformable-detr + dinodetr suites 70 passed / 5 skipped. Still owed before merge: the GPU run of `test_hub_matches_portable_on_cuda` plus an RF-DETR predict/train-step comparison vs `LIBREYOLO_HUB_KERNELS=0` — now meaningful since the kernel actually engages.
